File: apps/ios/TieComs/Core/SocketIOProtocol.swift
import Foundation

/// Protocolo Engine.IO v4 + Socket.IO v5 sobre WebSocket (solo texto, espacio de nombres "/").
///
/// Trama WebSocket = paquete Engine.IO: un dígito de tipo + datos.
///   0 open · 1 close · 2 ping · 3 pong · 4 message · 5 upgrade · 6 noop
/// Dentro de un "message" (4) va un paquete Socket.IO: tipo, espacio de nombres
/// opcional ("/x,"), id de ACK opcional (dígitos) y JSON.
///   0 CONNECT · 1 DISCONNECT · 2 EVENT · 3 ACK · 4 CONNECT_ERROR · 5/6 binarios (no usados)
enum EnginePacket: Equatable {
    case open(OpenInfo)
    case close
    case ping
    case pong
    case message(SocketPacket)
    case upgrade
    case noop
    case invalid(String)

    struct OpenInfo: Equatable {
        var sid: String
        var pingInterval: Int
        var pingTimeout: Int
        var maxPayload: Int
    }

    static func parse(_ text: String) -> EnginePacket {
        guard let first = text.first else { return .invalid(text) }
        let rest = String(text.dropFirst())
        switch first {
        case "0":
            guard let data = rest.data(using: .utf8),
                  let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let sid = obj["sid"] as? String else { return .invalid(text) }
            return .open(OpenInfo(
                sid: sid,
                pingInterval: (obj["pingInterval"] as? NSNumber)?.intValue ?? 25_000,
                pingTimeout: (obj["pingTimeout"] as? NSNumber)?.intValue ?? 20_000,
                maxPayload: (obj["maxPayload"] as? NSNumber)?.intValue ?? 1_000_000))
        case "1": return .close
        case "2": return .ping
        case "3": return .pong
        case "4":
            guard let p = SocketPacket.parse(rest) else { return .invalid(text) }
            return .message(p)
        case "5": return .upgrade
        case "6": return .noop
        default: return .invalid(text)
        }
    }

    var encoded: String {
        switch self {
        case .open: return "0"
        case .close: return "1"
        case .ping: return "2"
        case .pong: return "3"
        case .message(let p): return "4" + p.encoded
        case .upgrade: return "5"
        case .noop: return "6"
        case .invalid(let s): return s
        }
    }
}

struct SocketPacket: Equatable {
    enum Kind: Int { case connect = 0, disconnect = 1, event = 2, ack = 3, connectError = 4, binaryEvent = 5, binaryAck = 6 }

    var kind: Kind
    var namespace: String = "/"
    var ackId: Int?
    /// JSON crudo del cuerpo (objeto para CONNECT/CONNECT_ERROR, arreglo para EVENT/ACK).
    var json: String?

    static func parse(_ s: String) -> SocketPacket? {
        var i = s.startIndex
        guard i < s.endIndex, let t = s[i].wholeNumberValue, let kind = Kind(rawValue: t) else { return nil }
        i = s.index(after: i)
        // Paquetes binarios: "<n>-" adjuntos. No se usan; se descarta el prefijo.
        if kind == .binaryEvent || kind == .binaryAck {
            while i < s.endIndex, s[i] != "-" { i = s.index(after: i) }
            if i < s.endIndex { i = s.index(after: i) }
        }
        var nsp = "/"
        if i < s.endIndex, s[i] == "/" {
            guard let comma = s[i...].firstIndex(of: ",") else { return nil }
            nsp = String(s[i..<comma])
            i = s.index(after: comma)
        }
        var digits = ""
        while i < s.endIndex, let _ = s[i].wholeNumberValue { digits.append(s[i]); i = s.index(after: i) }
        let body = String(s[i...])
        return SocketPacket(kind: kind, namespace: nsp, ackId: digits.isEmpty ? nil : Int(digits), json: body.isEmpty ? nil : body)
    }

    var encoded: String {
        var out = String(kind.rawValue)
        if namespace != "/" { out += namespace + "," }
        if let ackId { out += String(ackId) }
        if let json { out += json }
        return out
    }

    /// EVENT: nombre y argumentos.
    var event: (name: String, args: [Any])? {
        guard kind == .event || kind == .binaryEvent, let arr = jsonArray, let name = arr.first as? String else { return nil }
        return (name, Array(arr.dropFirst()))
    }

    var jsonArray: [Any]? {
        guard let json, let data = json.data(using: .utf8) else { return nil }
        return (try? JSONSerialization.jsonObject(with: data, options: [.fragmentsAllowed])) as? [Any]
    }

    var jsonObject: [String: Any]? {
        guard let json, let data = json.data(using: .utf8) else { return nil }
        return (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
    }

    // MARK: Constructores

    static func connect(auth: [String: Any]) -> SocketPacket {
        SocketPacket(kind: .connect, json: jsonString(auth))
    }

    static func event(_ name: String, _ payload: Any?, ackId: Int? = nil) -> SocketPacket {
        var arr: [Any] = [name]
        if let payload { arr.append(payload) }
        return SocketPacket(kind: .event, ackId: ackId, json: jsonString(arr))
    }

    static func jsonString(_ obj: Any) -> String {
        guard JSONSerialization.isValidJSONObject(obj),
              let data = try? JSONSerialization.data(withJSONObject: obj, options: [.withoutEscapingSlashes]) else { return "null" }
        return String(decoding: data, as: UTF8.self)
    }
}

/// Convierte un valor JSON ya parseado en un DTO Decodable (tolerante).
enum JSONBridge {
    static func decode<T: Decodable>(_ type: T.Type, from value: Any?) -> T? {
        guard let value, JSONSerialization.isValidJSONObject(value),
              let data = try? JSONSerialization.data(withJSONObject: value) else { return nil }
        return try? JSONDecoder().decode(T.self, from: data)
    }
}

File: apps/android/app/src/main/java/com/tiecoms/app/core/SocketIoProtocol.kt
package com.tiecoms.app.core

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.longOrNull

/**
 * Engine.IO v4 (transporte websocket, solo texto). Cada frame empieza por el tipo:
 * 0 open · 1 close · 2 ping · 3 pong · 4 message · 5 upgrade · 6 noop.
 */
sealed interface EnginePacket {
    data class Open(val sid: String, val pingInterval: Long, val pingTimeout: Long) : EnginePacket
    data object Close : EnginePacket
    data object Ping : EnginePacket
    data object Pong : EnginePacket
    data class Message(val data: String) : EnginePacket
    data object Noop : EnginePacket
    data class Invalid(val raw: String) : EnginePacket
}

object EngineIo {
    const val PONG = "3"

    fun decode(frame: String): EnginePacket {
        if (frame.isEmpty()) return EnginePacket.Invalid(frame)
        val rest = frame.substring(1)
        return when (frame[0]) {
            '0' -> runCatching {
                val o = TcJson.parseToJsonElement(rest) as JsonObject
                EnginePacket.Open(
                    sid = (o["sid"] as? JsonPrimitive)?.contentOrNull ?: "",
                    pingInterval = (o["pingInterval"] as? JsonPrimitive)?.longOrNull ?: 25_000,
                    pingTimeout = (o["pingTimeout"] as? JsonPrimitive)?.longOrNull ?: 20_000,
                )
            }.getOrElse { EnginePacket.Invalid(frame) }
            '1' -> EnginePacket.Close
            '2' -> EnginePacket.Ping
            '3' -> EnginePacket.Pong
            '4' -> EnginePacket.Message(rest)
            '6' -> EnginePacket.Noop
            else -> EnginePacket.Invalid(frame)
        }
    }

    fun message(socketIoPacket: String) = "4$socketIoPacket"
}

/**
 * Socket.IO v5 sobre Engine.IO: `<tipo>[<nsp>,][<ackId>][<json>]`.
 * 0 CONNECT · 1 DISCONNECT · 2 EVENT · 3 ACK · 4 CONNECT_ERROR · 5/6 binarios (no usados).
 */
sealed interface SioPacket {
    data class Connect(val sid: String?) : SioPacket
    data object Disconnect : SioPacket
    data class Event(val name: String, val args: List<JsonElement>, val ackId: Long?) : SioPacket
    data class Ack(val id: Long, val args: List<JsonElement>) : SioPacket
    data class ConnectError(val message: String) : SioPacket
    data class Unsupported(val raw: String) : SioPacket
}

object SocketIo {
    fun decode(packet: String): SioPacket {
        if (packet.isEmpty()) return SioPacket.Unsupported(packet)
        val type = packet[0]
        var i = 1
        if (type == '5' || type == '6') return SioPacket.Unsupported(packet)
        // Espacio de nombres distinto de "/": "/admin,..."
        if (i < packet.length && packet[i] == '/') {
            val comma = packet.indexOf(',', i)
            if (comma < 0) return SioPacket.Unsupported(packet)
            i = comma + 1
        }
        val idStart = i
        while (i < packet.length && packet[i].isDigit()) i++
        val ackId = if (i > idStart) packet.substring(idStart, i).toLongOrNull() else null
        val json = packet.substring(i)
        val data: JsonElement? = if (json.isEmpty()) null else runCatching { TcJson.parseToJsonElement(json) }.getOrNull()
        return when (type) {
            '0' -> SioPacket.Connect(((data as? JsonObject)?.get("sid") as? JsonPrimitive)?.contentOrNull)
            '1' -> SioPacket.Disconnect
            '2' -> {
                val arr = data as? JsonArray ?: return SioPacket.Unsupported(packet)
                val name = (arr.firstOrNull() as? JsonPrimitive)?.contentOrNull ?: return SioPacket.Unsupported(packet)
                SioPacket.Event(name, arr.drop(1), ackId)
            }
            '3' -> {
                val arr = data as? JsonArray ?: return SioPacket.Unsupported(packet)
                if (ackId == null) SioPacket.Unsupported(packet) else SioPacket.Ack(ackId, arr.toList())
            }
            '4' -> SioPacket.ConnectError(
                ((data as? JsonObject)?.get("message") as? JsonPrimitive)?.contentOrNull
                    ?: (data as? JsonPrimitive)?.contentOrNull ?: "error",
            )
            else -> SioPacket.Unsupported(packet)
        }
    }

    fun connect(auth: JsonObject): String = "0$auth"

    fun event(name: String, data: JsonElement?, ackId: Long? = null): String {
        val arr = buildJsonArray {
            add(JsonPrimitive(name))
            if (data != null) add(data)
        }
        return "2${ackId ?: ""}$arr"
    }
}
